File: go/git/skill.go
// Package main implements git-skill — Git operations via shell_exec.
//
// @skill:id      ai.luminarys.go.git
// @skill:name    "Git Skill"
// @skill:version 1.0.0
// @skill:desc    "Git repository operations: status, diff, log, blame, commit, branch management."
//
// @skill:require shell git **
//
//go:generate lmsk -lang go -verbose .
package main

import (
	"fmt"
	"strings"

	sdk "github.com/LuminarysAI/sdk-go"
)

// ── helpers ───────────────────────────────────────────────────────────────────

func git(ctx *sdk.Context, workdir string, args ...string) (sdk.ShellExecResult, error) {
	cmd := "git " + strings.Join(args, " ")
	return sdk.ShellExec(sdk.ShellExecRequest{
		Command:   cmd,
		Workdir:   workdir,
		TimeoutMs: 30000,
	})
}

func gitOutput(ctx *sdk.Context, workdir string, args ...string) (string, error) {
	result, err := git(ctx, workdir, args...)
	if err != nil {
		return "", err
	}
	if result.ExitCode != 0 {
		msg := strings.TrimSpace(result.Output)
		return "", fmt.Errorf("git %s: exit %d: %s", args[0], result.ExitCode, msg)
	}
	return result.Output, nil
}

// ── Read-only operations ──────────────────────────────────────────────────────

// Init initializes a new git repository in the given directory.
// @skill:method init "Initialize a new git repository."
// @skill:param  workdir required "Absolute path to the directory to initialize"
// @skill:result "Initialization confirmation"
func Init(ctx *sdk.Context, workdir string) (string, error) {
	return gitOutput(ctx, workdir, "init")
}

// Status returns the working tree status (porcelain v2 format).
// @skill:method status "Show working tree status: staged, unstaged, and untracked files."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:result "Porcelain status output, one file per line"
func Status(ctx *sdk.Context, workdir string) (string, error) {
	return gitOutput(ctx, workdir, "status", "--porcelain=v2", "--branch")
}

// Diff returns the diff of unstaged changes, or staged changes if staged=true.
// @skill:method diff "Show file changes as unified diff."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  path    optional "Limit diff to this file or directory"
// @skill:param  staged  optional "Show staged (cached) changes instead of unstaged" default:false
// @skill:result "Unified diff output"
func Diff(ctx *sdk.Context, workdir, path string, staged bool) (string, error) {
	args := []string{"diff", "--no-color"}
	if staged {
		args = append(args, "--cached")
	}
	if path != "" {
		args = append(args, "--", path)
	}
	return gitOutput(ctx, workdir, args...)
}

// Log returns recent commit history.
// @skill:method log "Show commit history."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  count   optional "Number of commits to show" default:20
// @skill:param  path    optional "Limit log to this file or directory"
// @skill:param  oneline optional "One line per commit (hash + subject)" default:false
// @skill:result "Git log output"
func Log(ctx *sdk.Context, workdir string, count int64, path string, oneline bool) (string, error) {
	n := count
	if n <= 0 {
		n = 20
	}
	args := []string{"log", fmt.Sprintf("-n%d", n), "--no-color"}
	if oneline {
		args = append(args, "--oneline")
	} else {
		args = append(args, "--format=format:%H %ai %an <%ae>%n%s%n")
	}
	if path != "" {
		args = append(args, "--", path)
	}
	return gitOutput(ctx, workdir, args...)
}

// Show displays the contents of a commit.
// @skill:method show "Show commit details and diff."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  ref     optional "Commit hash, tag, or branch name" default:HEAD
// @skill:param  stat    optional "Show diffstat only (no full diff)" default:false
// @skill:result "Commit details and diff"
func Show(ctx *sdk.Context, workdir, ref string, stat bool) (string, error) {
	if ref == "" {
		ref = "HEAD"
	}
	args := []string{"show", "--no-color", ref}
	if stat {
		args = []string{"show", "--no-color", "--stat", ref}
	}
	return gitOutput(ctx, workdir, args...)
}

// Blame shows line-by-line authorship for a file.
// @skill:method blame "Show who last modified each line of a file."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  path    required "File path relative to repo root"
// @skill:result "Blame output with commit, author, and line content"
func Blame(ctx *sdk.Context, workdir, path string) (string, error) {
	return gitOutput(ctx, workdir, "blame", "--no-color", "--", path)
}

// Branches lists all branches (local and remote).
// @skill:method branches "List all branches."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:result "Branch list with current branch marked"
func Branches(ctx *sdk.Context, workdir string) (string, error) {
	return gitOutput(ctx, workdir, "branch", "-a", "--no-color")
}

// Tags lists all tags.
// @skill:method tags "List all tags."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:result "Tag list"
func Tags(ctx *sdk.Context, workdir string) (string, error) {
	return gitOutput(ctx, workdir, "tag", "-l")
}

// DiffStat returns diffstat summary (files changed, insertions, deletions).
// @skill:method diff_stat "Show summary of changes between two refs."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  from    optional "Start ref (commit, branch, tag)" default:HEAD~1
// @skill:param  to      optional "End ref" default:HEAD
// @skill:result "Diffstat summary"
func DiffStat(ctx *sdk.Context, workdir, from, to string) (string, error) {
	if from == "" {
		from = "HEAD~1"
	}
	if to == "" {
		to = "HEAD"
	}
	return gitOutput(ctx, workdir, "diff", "--stat", "--no-color", from+"..."+to)
}

// ── Write operations ──────────────────────────────────────────────────────────

// Add stages files for commit.
// @skill:method add "Stage files for the next commit."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  paths   required "Space-separated file paths to stage (use . for all)"
// @skill:result "Status after staging"
func Add(ctx *sdk.Context, workdir, paths string) (string, error) {
	args := append([]string{"add"}, strings.Fields(paths)...)
	if _, err := gitOutput(ctx, workdir, args...); err != nil {
		return "", err
	}
	return gitOutput(ctx, workdir, "status", "--porcelain=v2", "--branch")
}

// Commit creates a new commit with the given message.
// @skill:method commit "Create a commit with staged changes."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  message required "Commit message"
// @skill:result "Commit hash and summary"
func Commit(ctx *sdk.Context, workdir, message string) (string, error) {
	return gitOutput(ctx, workdir, "commit", "-m", fmt.Sprintf("%q", message))
}

// CreateBranch creates and switches to a new branch.
// @skill:method create_branch "Create a new branch and switch to it."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  name    required "New branch name"
// @skill:param  from    optional "Base ref to branch from" default:HEAD
// @skill:result "Confirmation message"
func CreateBranch(ctx *sdk.Context, workdir, name, from string) (string, error) {
	args := []string{"checkout", "-b", name}
	if from != "" {
		args = append(args, from)
	}
	return gitOutput(ctx, workdir, args...)
}

// Checkout switches to an existing branch or ref.
// @skill:method checkout "Switch to a branch, tag, or commit."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  ref     required "Branch name, tag, or commit hash"
// @skill:result "Confirmation message"
func Checkout(ctx *sdk.Context, workdir, ref string) (string, error) {
	return gitOutput(ctx, workdir, "checkout", ref)
}

// Stash saves uncommitted changes to the stash.
// @skill:method stash "Stash uncommitted changes."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  message optional "Stash message"
// @skill:result "Stash confirmation"
func Stash(ctx *sdk.Context, workdir, message string) (string, error) {
	args := []string{"stash", "push"}
	if message != "" {
		args = append(args, "-m", message)
	}
	return gitOutput(ctx, workdir, args...)
}

// StashPop restores the most recent stash.
// @skill:method stash_pop "Restore the most recently stashed changes."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:result "Restored changes summary"
func StashPop(ctx *sdk.Context, workdir string) (string, error) {
	return gitOutput(ctx, workdir, "stash", "pop")
}

// StashList shows all stash entries.
// @skill:method stash_list "List all stash entries."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:result "Stash entries"
func StashList(ctx *sdk.Context, workdir string) (string, error) {
	return gitOutput(ctx, workdir, "stash", "list")
}

// Restore discards changes in a file (resets to HEAD).
// @skill:method restore "Discard uncommitted changes in a file."
// @skill:param  workdir required "Absolute path to the git repository"
// @skill:param  path    required "File path to restore"
// @skill:result "Confirmation"
func Restore(ctx *sdk.Context, workdir, path string) (string, error) {
	_, err := gitOutput(ctx, workdir, "checkout", "--", path)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("restored %s", path), nil
}

File: rust/git/src/skill.rs
//! git-skill — Git repository operations via shell_exec (Rust).
//!
//! @skill:id      ai.luminarys.rust.git
//! @skill:name    "Git Skill (Rust)"
//! @skill:version 1.0.0
//! @skill:desc    "Git repository operations: status, diff, log, blame, commit, branch management."
//!
//! @skill:require shell git **

use luminarys_sdk::prelude::*;

// ── helpers ───────────────────────────────────────────────────────────────────

fn git(workdir: &str, args: &[&str]) -> Result<String, SkillError> {
    let cmd = format!("git {}", args.join(" "));
    let result = shell_exec(&ShellExecRequest {
        command: cmd.clone(),
        workdir: workdir.into(),
        timeout_ms: 30000,
        tail: 0,
        grep: String::new(),
        as_daemon: false,
        log_file: String::new(),
    })?;
    if result.exit_code != 0 {
        let msg = result.output.trim().to_string();
        return Err(SkillError(format!(
            "git {}: exit {}: {}",
            args[0], result.exit_code, msg
        )));
    }
    Ok(result.output)
}

// ── Read-only operations ──────────────────────────────────────────────────────

/// Initialize a new git repository.
/// @skill:method init "Initialize a new git repository."
/// @skill:param  workdir required "Absolute path to the directory to initialize"
/// @skill:result "Initialization confirmation"
pub fn init(_ctx: &mut Context, workdir: String) -> Result<String, SkillError> {
    git(&workdir, &["init"])
}

/// Show working tree status.
/// @skill:method status "Show working tree status: staged, unstaged, and untracked files."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:result "Porcelain status output"
pub fn status(_ctx: &mut Context, workdir: String) -> Result<String, SkillError> {
    git(&workdir, &["status", "--porcelain=v2", "--branch"])
}

/// Show file changes as unified diff.
/// @skill:method diff "Show file changes as unified diff."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  path    optional "Limit diff to this file or directory"
/// @skill:param  staged  optional "Show staged changes instead of unstaged"
/// @skill:result "Unified diff output"
pub fn diff(_ctx: &mut Context, workdir: String, path: String, staged: bool) -> Result<String, SkillError> {
    let mut args = vec!["diff", "--no-color"];
    if staged {
        args.push("--cached");
    }
    if !path.is_empty() {
        args.push("--");
        args.push(&path);
    }
    git(&workdir, &args)
}

/// Show commit history.
/// @skill:method log "Show commit history."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  count   optional "Number of commits to show"
/// @skill:param  path    optional "Limit log to this file or directory"
/// @skill:param  oneline optional "One line per commit"
/// @skill:result "Git log output"
pub fn log(_ctx: &mut Context, workdir: String, count: i64, path: String, oneline: bool) -> Result<String, SkillError> {
    let n = if count <= 0 { 20 } else { count };
    let n_arg = format!("-n{}", n);
    let mut args = vec!["log", &n_arg, "--no-color"];
    if oneline {
        args.push("--oneline");
    } else {
        args.push("--format=format:%H %ai %an <%ae>%n%s%n");
    }
    if !path.is_empty() {
        args.push("--");
        args.push(&path);
    }
    git(&workdir, &args)
}

/// Show commit details and diff.
/// @skill:method show "Show commit details and diff."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  ref     optional "Commit hash, tag, or branch name"
/// @skill:param  stat    optional "Show diffstat only"
/// @skill:result "Commit details and diff"
pub fn show(_ctx: &mut Context, workdir: String, r#ref: String, stat: bool) -> Result<String, SkillError> {
    let ref_val = if r#ref.is_empty() { "HEAD" } else { &r#ref };
    let mut args = vec!["show", "--no-color"];
    if stat {
        args.push("--stat");
    }
    args.push(ref_val);
    git(&workdir, &args)
}

/// Show who last modified each line of a file.
/// @skill:method blame "Show who last modified each line of a file."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  path    required "File path relative to repo root"
/// @skill:result "Blame output"
pub fn blame(_ctx: &mut Context, workdir: String, path: String) -> Result<String, SkillError> {
    git(&workdir, &["blame", "--no-color", "--", &path])
}

/// List all branches.
/// @skill:method branches "List all branches."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:result "Branch list"
pub fn branches(_ctx: &mut Context, workdir: String) -> Result<String, SkillError> {
    git(&workdir, &["branch", "-a", "--no-color"])
}

/// List all tags.
/// @skill:method tags "List all tags."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:result "Tag list"
pub fn tags(_ctx: &mut Context, workdir: String) -> Result<String, SkillError> {
    git(&workdir, &["tag", "-l"])
}

/// Show summary of changes between two refs.
/// @skill:method diff_stat "Show summary of changes between two refs."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  from    optional "Start ref"
/// @skill:param  to      optional "End ref"
/// @skill:result "Diffstat summary"
pub fn diff_stat(_ctx: &mut Context, workdir: String, from: String, to: String) -> Result<String, SkillError> {
    let f = if from.is_empty() { "HEAD~1".to_string() } else { from };
    let t = if to.is_empty() { "HEAD".to_string() } else { to };
    let range = format!("{}...{}", f, t);
    git(&workdir, &["diff", "--stat", "--no-color", &range])
}

// ── Write operations ──────────────────────────────────────────────────────────

/// Stage files for the next commit.
/// @skill:method add "Stage files for the next commit."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  paths   required "Space-separated file paths to stage (use . for all)"
/// @skill:result "Status after staging"
pub fn add(_ctx: &mut Context, workdir: String, paths: String) -> Result<String, SkillError> {
    let parts: Vec<&str> = paths.split_whitespace().collect();
    let mut args = vec!["add"];
    args.extend(parts);
    git(&workdir, &args)?;
    git(&workdir, &["status", "--porcelain=v2", "--branch"])
}

/// Create a commit with staged changes.
/// @skill:method commit "Create a commit with staged changes."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  message required "Commit message"
/// @skill:result "Commit hash and summary"
pub fn commit(_ctx: &mut Context, workdir: String, message: String) -> Result<String, SkillError> {
    let quoted = format!("\"{}\"", message.replace('"', "\\\""));
    git(&workdir, &["commit", "-m", &quoted])
}

/// Create a new branch and switch to it.
/// @skill:method create_branch "Create a new branch and switch to it."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  name    required "New branch name"
/// @skill:param  from    optional "Base ref to branch from"
/// @skill:result "Confirmation"
pub fn create_branch(_ctx: &mut Context, workdir: String, name: String, from: String) -> Result<String, SkillError> {
    let mut args = vec!["checkout", "-b", &name];
    if !from.is_empty() {
        args.push(&from);
    }
    git(&workdir, &args)
}

/// Switch to a branch, tag, or commit.
/// @skill:method checkout "Switch to a branch, tag, or commit."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  ref     required "Branch name, tag, or commit hash"
/// @skill:result "Confirmation"
pub fn checkout(_ctx: &mut Context, workdir: String, r#ref: String) -> Result<String, SkillError> {
    git(&workdir, &["checkout", &r#ref])
}

/// Stash uncommitted changes.
/// @skill:method stash "Stash uncommitted changes."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  message optional "Stash message"
/// @skill:result "Stash confirmation"
pub fn stash(_ctx: &mut Context, workdir: String, message: String) -> Result<String, SkillError> {
    let mut args = vec!["stash", "push"];
    if !message.is_empty() {
        args.push("-m");
        args.push(&message);
    }
    git(&workdir, &args)
}

/// Restore the most recently stashed changes.
/// @skill:method stash_pop "Restore the most recently stashed changes."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:result "Restored changes"
pub fn stash_pop(_ctx: &mut Context, workdir: String) -> Result<String, SkillError> {
    git(&workdir, &["stash", "pop"])
}

/// List all stash entries.
/// @skill:method stash_list "List all stash entries."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:result "Stash entries"
pub fn stash_list(_ctx: &mut Context, workdir: String) -> Result<String, SkillError> {
    git(&workdir, &["stash", "list"])
}

/// Discard uncommitted changes in a file.
/// @skill:method restore "Discard uncommitted changes in a file."
/// @skill:param  workdir required "Absolute path to the git repository"
/// @skill:param  path    required "File path to restore"
/// @skill:result "Confirmation"
pub fn restore(_ctx: &mut Context, workdir: String, path: String) -> Result<String, SkillError> {
    git(&workdir, &["checkout", "--", &path])?;
    Ok(format!("restored {}", path))
}
